File: infer/tests/codetoanalyze/java/pulse/taint/Exceptions.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

class Exceptions {

  static native void mayExcept() throws Exception;

  public static void FN_sinkInCatchBad1() {
    Object source = InferTaint.inferSecretSource();
    try {
      mayExcept();
    } catch (Exception e) {
      InferTaint.inferSensitiveSink(source);
    }
  }

  public static void FN_sinkInCatchBad2() {
    Object source = null;
    try {
      source = InferTaint.inferSecretSource();
      mayExcept();
    } catch (Exception e) {
      InferTaint.inferSensitiveSink(source);
    }
  }

  public static void FN_sinkAfterCatchBad() {
    Object source = InferTaint.inferSecretSource();
    try {
      mayExcept();
      source = null;
    } catch (Exception e) {
    }
    InferTaint.inferSensitiveSink(source);
  }

  public static void sinkAfterCatchOk() {
    Object source = InferTaint.inferSecretSource();
    try {
      mayExcept();
      source = null;
    } catch (Exception e) {
      source = null;
    }
    InferTaint.inferSensitiveSink(source);
  }

  public static void sinkInFinallyBad1() throws Exception {
    Object source = InferTaint.inferSecretSource();
    try {
      mayExcept();
    } finally {
      InferTaint.inferSensitiveSink(source);
    }
  }

  public static void sinkInFinallyBad2() throws Exception {
    Object source = null;
    try {
      mayExcept();
      source = InferTaint.inferSecretSource();
    } finally {
      InferTaint.inferSensitiveSink(source);
    }
  }

  public static void FN_sinkInFinallyBad3() {
    Object source = null;
    try {
      mayExcept();
    } catch (Exception e) {
      source = InferTaint.inferSecretSource();
    } finally {
      InferTaint.inferSensitiveSink(source);
    }
  }

  public static void sinkAfterFinallyOk1() throws Exception {
    Object source = InferTaint.inferSecretSource();
    try {
      mayExcept();
    } finally {
      source = null;
    }
    InferTaint.inferSensitiveSink(source);
  }

  public static void sinkAfterFinallyOk2() {
    Object source = null;
    try {
      mayExcept();
      source = InferTaint.inferSecretSource();
    } catch (Exception e) {
      source = InferTaint.inferSecretSource();
    } finally {
      source = null;
    }
    InferTaint.inferSensitiveSink(source);
  }

  public static void callSinkThenThrow(Object param) throws Exception {
    InferTaint.inferSensitiveSink(param);
    throw new Exception();
  }

  public static void callSinkThenThrowBad() throws Exception {
    callSinkThenThrow(InferTaint.inferSecretSource());
  }

  public static void doThrow(Object param) throws RuntimeException {
    throw new RuntimeException(param.toString());
  }

  // FN because object initialization of the exception thrown in doThrow(), which is
  // [java.lang.RuntimeException.<init>], is treated as an unknown function. The unknown function
  // heuristic is defeated by this form of call. We should revisit the unknown function heuristic to
  // be more like the taint unknown function heuristic and detect assignment via the first argument,
  // or just model [<init>] unknown functions separately
  public static void FN_callSinkWithSourceInsideExceptionObjectBad() {
    try {
      doThrow(InferTaint.inferSecretSource());
    } catch (RuntimeException e) {
      InferTaint.inferSensitiveSink(e);
    }
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/Exceptions.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

import codetoanalyze.kotlin.pulse.InferTaint.inferSecretSource
import codetoanalyze.kotlin.pulse.InferTaint.inferSensitiveSink

class Exceptions {

  @Throws(Exception::class) external fun mayExcept()

  fun FN_sinkInCatchBad1() {
    val source = inferSecretSource()
    try {
      mayExcept()
    } catch (e: Exception) {
      inferSensitiveSink(source)
    }
  }

  fun FN_sinkInCatchBad2() {
    var source: Any? = null
    try {
      source = inferSecretSource()
      mayExcept()
    } catch (e: Exception) {
      inferSensitiveSink(source)
    }
  }

  fun FN_sinkAfterCatchBad() {
    var source: Any? = inferSecretSource()
    try {
      mayExcept()
      source = null
    } catch (e: Exception) {}
    inferSensitiveSink(source)
  }

  fun sinkAfterCatchOk() {
    var source: Any? = inferSecretSource()
    source =
        try {
          mayExcept()
          null
        } catch (e: Exception) {
          null
        }
    inferSensitiveSink(source)
  }

  @Throws(Exception::class)
  fun sinkInFinallyBad1() {
    val source = inferSecretSource()
    try {
      mayExcept()
    } finally {
      inferSensitiveSink(source)
    }
  }

  @Throws(Exception::class)
  fun sinkInFinallyBad2() {
    var source: Any? = null
    try {
      mayExcept()
      source = inferSecretSource()
    } finally {
      inferSensitiveSink(source)
    }
  }

  fun FN_sinkInFinallyBad3() {
    var source: Any? = null
    try {
      mayExcept()
    } catch (e: Exception) {
      source = inferSecretSource()
    } finally {
      inferSensitiveSink(source)
    }
  }

  @Throws(Exception::class)
  fun sinkAfterFinallyOk1() {
    var source: Any? = inferSecretSource()
    try {
      mayExcept()
    } finally {
      source = null
    }
    inferSensitiveSink(source)
  }

  fun sinkAfterFinallyOk2() {
    var source: Any? = null
    try {
      mayExcept()
      source = inferSecretSource()
    } catch (e: Exception) {
      source = inferSecretSource()
    } finally {
      source = null
    }
    inferSensitiveSink(source)
  }

  @Throws(Exception::class)
  fun callSinkThenThrow(param: Any?) {
    inferSensitiveSink(param)
    throw Exception()
  }

  @Throws(Exception::class)
  fun callSinkThenThrowBad() {
    callSinkThenThrow(inferSecretSource())
  }

  @Throws(RuntimeException::class)
  fun doThrow(param: Any) {
    throw RuntimeException(param.toString())
  }

  // see comment on the similar function in codetoanalyze/java/pulse/taint/Exceptions.java
  fun FN_callSinkWithSourceInsideExceptionObjectBad() {
    try {
      doThrow(inferSecretSource())
    } catch (e: RuntimeException) {
      inferSensitiveSink(e)
    }
  }
}
